### PR TITLE
[Enhancement 3.2 3.3] COT, TOT strategy for engineer role

### PR DESCRIPTION
**Pull request by LeonZh0u**
WIP implemented COT, will push TOT if the current way of calling aask() makes sense

### CODE REVIEW

File: examples/code_generation.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+from metagpt.roles.engineer import Engineer
+from metagpt.utils.common import CodeParser
+from mock import (
+    TASKS,
+    MockMessages,
+)
+
+import asyncio
+
+@pytest.mark.asyncio
```
Comment (geekan) [inline examples/code_generation.py:16]: PEP8

File: examples/code_generation.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
```
Comment (geekan) [inline examples/code_generation.py:2]: 加上合适的带作者的文件头

File: metagpt/actions/action.py [hunk 3]
```diff
@@ -39,11 +41,20 @@ def __str__(self):
     def __repr__(self):
         return self.__str__()
 
-    async def _aask(self, prompt: str, system_msgs: Optional[list[str]] = None) -> str:
+    async def _aask(self, prompt: str, strategy: PromptStrategyType = None, task: Task = None, system_msgs: Optional[list[str]] = None) -> str:
```
Comment (geekan) [inline metagpt/actions/action.py:44]: 使用Optional装饰可能None的字段

File: metagpt/actions/action.py [hunk 4]
```diff
@@ -39,11 +41,20 @@ def __str__(self):
...
             system_msgs = []
         system_msgs.append(self.prefix)
+
+        if strategy == PromptStrategyType.CHAIN_OF_THOUGHT:
+            for PROMPT, output_key in zip(task.prompts, task.task_output_keys):
+                PROMPT = PROMPT.format(**task.task_args_pool)
+                task.task_args_pool[output_key] = await self._aask(PROMPT)
+                prompt = PROMPT
+        elif strategy == PromptStrategyType.TREE_OF_THOUGHT:
+            pass
```
Comment (geekan) [inline metagpt/actions/action.py:56]: 这里应该要raise SomeException

File: metagpt/actions/task.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+from typing import Dict
```
Comment (geekan) [inline metagpt/actions/task.py:1]: 需要合适的文件头

File: metagpt/actions/write_code.py [hunk 6]
```diff
@@ -41,6 +44,47 @@
...
+-----
+# Context
+{context}
+-----
+NOTICE
+Role: You are a professional engineer; Given the previous context, the main goal is to create a tutorial-like step by step plan for writing {filename} in Python 3.9 code (but you can also use other programming language)
+ATTENTION: Use '##' to SPLIT SECTIONS, not '#'. Output format carefully referenced "Format example".
+-----
+## Step by Step Implementation Plan: {filename}
+"""
```
Comment (geekan) [inline metagpt/actions/write_code.py:57]: 我比较建议如果是写代码，建议这里就直接让它写伪代码

Comment (geekan) [reply metagpt/actions/write_code.py:57]: 并且在伪代码上加合适的注释，或者写多份伪代码

File: metagpt/actions/write_code.py [hunk 7]
```diff
@@ -65,17 +109,41 @@ def _save(self, context, filename, code):
...
+                        # a pool of key value pairs to format the prompts
+                        task_args_pool = {
+                            "context": context, 
+                            "filename": filename, 
+                            # intermediate result is empty
+                            "plan": ""
+                        },
+                        # an ordered list of prompt output keys
+                        task_output_keys = ["plan", "code"])
+            code = await self.write_code('', strategy = strategy, task = task)
```
Comment (geekan) [inline metagpt/actions/write_code.py:139]: 没有处理TOT等其他状态

File: metagpt/strategy/cot.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
+class ChainOfThoughtBase(object):
```
Comment (weburnit) [inline metagpt/strategy/cot.py:1]: from abc import ABC
class ChainOfThoughtBase(ABC):

**Comment by geekan** [comment]
Please resolve all conflicts and Review comments @LeonZh0u

**Comment by geekan** [comment]
This PR has been closed because it took too long, feel free to reopen it.

**Pull request closed without merging**